rfcs/2021-09-14-3910-schema-support.md

```diff
@@ -0,0 +1,472 @@
+remove erroneously delivered events, without having to re-process the entire
+event stream again.
+
+Event schemas are one way of enforcing the contract between teams, and ensuring
+that only valid events end up at their expected destination.
+
+## User Experience
+
+Vector operators should have the option to guarantee the shape of events based
+on event schemas.
+
+To do this, they can do one of two things.
+
+1. For sources with a **known** schema, Vector guarantees the schema of each
+   event received by that source, by implicitly enforcing the source-schema on
+   those events.
+
+   - [ ] **TODO** Which sources have a fixed/known schema?
+
+   - [ ] **TODO** Should we ever allow operators to disable and/or change the
+     schema of those sources?
+
+2. For sources that don't have a fixed schema defined, operators can define
+   their own schema — or use a schema pre-defined within Vector — to enforce the
+   shape of events.
+
+### Schema Validation
+
+Support for schema validation is imperative to the overarching goal of schema
+support in Vector. Without validation, there can be no guarantee that an event
+matches the expected schema, which in turn prevents other systems within Vector
+to rely on the information that an even adheres to a given schema.
+
+Schema validation is an expensive operation, however. For every incoming event,
+the entire shape of the event has to be compared against a pre-defined schema.
+While we can optimize this as much as possible, the inherent cost of walking an
+event and doing comparisons cannot be ignored, and has to be accepted as an
+inherent cost of implementing schema support.
+
+As developers, we instinctively search for solutions to this problem, such as
+sampled schema validation, to limit the impact on performance, but because all
+internal systems will be built _assuming_ schema adherence is guaranteed, there
+can be no room for error, nor for shortcuts that give you mostly statistically
+correct results, but not always.
+
+Even so, aside from the impact on the internal systems, if we promote strict
+schema adherence, the expectation is that no badly formatted events will end up
+at the configured destination.
+
+Because of this fact, this RFC proposes one solution to start with: you either
+_enable_ schema support, and accept the inherent performance cost, or you
+_disable_ support, cannot be guaranteed any schema adherence, but keep the most
+amount of performance. In the future, alternative "strategies" might be added,
+to find a middle-ground, but this would likely require a new RFC, given the
+mentioned implications.
+
+### Invalid Events
+
+For any source that has a schema configured, it can either **drop** or
+**accept** invalid events.
+
+#### Dropping Events
+
+When a source is configured to drop invalid events, and an event does not match
+the schema, the event won't be allowed to proceed to the next component in the
+pipeline.
```

Comment:
Does validation only happen at the `source` layer? Could a user `remap` events to break the schema contract before events reach the `sink` which is expecting a guaranteed shape?

EDIT: I believe this is answered by "merging" of schemas as events flow through the topology.